### ISSUE TITLE
False negative when updating React state with functional setState and object spread (prevState)

### ISSUE THREAD

**Issue by gabriel-h-nx** [microsoft/TypeScript#63041]
## Description

When using React state updates with the functional form of `setState`, TypeScript allows extra properties to be added to the returned object without reporting any type error. This results in a **false negative**, where invalid state keys silently pass type checking.

This commonly happens when spreading `prevState` and adding a property that does not exist in the state type.

## Example

```ts
type State = {
  states: string[];
  page: number;
};

const [filters, setFilters] = useState<State>({
  states: [],
  page: 1,
});

setFilters(prev => ({
  ...prev,
  state: "SP", // ❌ This property does NOT exist in `State`
}));
```

## Expected behavior

A type error should be reported when returning an object with properties that do not exist on the state type, even when using the functional setState(prev => ...) pattern.

## Environment

React with useState (this happens in versions: 18 and 19);

TypeScript (strict mode enabled);

ESLint v9 (Flat Config);

@typescript-eslint;

eslint-plugin-react-hooks;

TypeScript does not report any error in this scenario, even though state is not part of the State type.

## ESLint configuration

```ts
export default defineConfig([
  {
    files: ["**/*.{ts,tsx,js,jsx}"],
    plugins: {
      react: reactPlugin,
      "@typescript-eslint": typescriptPlugin,
      "react-hooks": reactHooksPlugin,
    },
    languageOptions: {
      parser: "@typescript-eslint/parser",
      parserOptions: {
        project: "./tsconfig.eslint.json",
      },
    },
    rules: {
      "react-hooks/rules-of-hooks": "error",
      "@typescript-eslint/no-unused-vars": "warn",
    },
  },
]);
```

Anybody can help me with this?

**Comment by mkantor** [comment]
I see you labeled your issue as a duplicate which maybe means you already found this, but unless you're proposing some specific extension of the [excess property check](https://www.typescriptlang.org/docs/handbook/2/objects.html#excess-property-checks) semantics, I think you're looking for #12936?

**Comment by MartinJohns** [comment]
Duplicate of #39998. Edit: nevermind, this is the wrong one. There was another issue about using the contextual type for the return type. The issue here is that the function passed to `setFilters` has no return type annotation, so the return type is inferred and then checked for compatibility. Objects are allowed to have additional properties.

> I see you labeled your issue as a duplicate which maybe means you already found this

@mkantor He just used the wrong issue template, where it gets added automatically.

**Comment by mkantor** [comment]
I don't think this is exactly #39998, because [there's also no error if the properties are written out rather than coming from a spread](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgVwM4FMDKMCGN1wC+cAZlBCHAORTo4DGMVAsAFBswCeYB2eBAXkRs4cVLnyoAXGJhRgAOwDmAbQC6AbhFwwOJehkLkIAEbooW1oUtt6EBeLgqSwADb4oqADRj0MAGJuHqhqcEJoWBLoADx8+AB8ABQI2uL80k5qXtq6+jIAjNlWAJQ2rBgBQeaoiWC0AG5h8XDJqVEZ6kWiuQZwhW38MgBEmAAKQz4A9JNwgDLkcAAqABbAqDrkPLCccAAmEOhrAHIA8gtw6AAeq-CKcAAGceh3bITFpUA).

[Here's](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAogHgQwLZgDbQLxQBQEooYB8UA3lAgFxQDOwATgJYB2A5lAL4CwAUDwMYB7JrShMBMOnQF0AEhDoQq8ZGkw58RHCR5RyVAOT6ANDqgQ4fCNWoAFKZDqgDY4OSZnJ04z3a4gA) a more minimal example (which also has no spread):

```ts
type Example = () => { a: string }

const noErrorHere: Example = () => ({
  a: '',
  excessProperty: 'not an error',
})
```

**Comment by gabriel-h-nx** [comment]
> I see you labeled your issue as a duplicate which maybe means you already found this, but unless you're proposing some specific extension of the [excess property check](https://www.typescriptlang.org/docs/handbook/2/objects.html#excess-property-checks) semantics, I think you're looking for [#12936](https://github.com/microsoft/TypeScript/issues/12936)?

No, Sorry about that. I marked this response incorrectly by mistake.

**Comment by gabriel-h-nx** [comment]
I will open another issue thanks
